pkg/k8s/watchers/cilium_cidr_group.go

```diff
@@ -0,0 +1,253 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package watchers
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	cilium_v2_alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	"github.com/cilium/cilium/pkg/k8s/watchers/resources"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/policy/api"
+)
+
+func (k *K8sWatcher) onUpsertCIDRGroup(
+	cidrGroup *cilium_v2_alpha1.CiliumCIDRGroup,
+	cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup,
+	cnpCache map[resource.Key]*types.SlimCNP,
+	cs client.Clientset,
+	apiGroup, metricLabel string,
+) error {
+	var (
+		equal  bool
+		action string
+	)
+
+	// wrap k.K8sEventReceived call into a naked func() to capture equal in the closure
+	defer func() {
+		k.K8sEventReceived(apiGroup, metricLabel, action, true, equal)
+	}()
+
+	oldCidrGroup, ok := cidrGroupCache[cidrGroup.Name]
+	if ok && oldCidrGroup.Spec.DeepEqual(&cidrGroup.Spec) {
+		equal = true
+		return nil
+	} else if ok {
+		action = resources.MetricUpdate
+	} else {
+		action = resources.MetricCreate
+	}
+
+	cidrGroupCpy := cidrGroup.DeepCopy()
+	cidrGroupCache[cidrGroup.Name] = cidrGroupCpy
+
+	err := k.updateCIDRGroupRefPolicies(cidrGroup.Name, cidrGroupCache, cnpCache, cs)
+
+	k.K8sEventProcessed(metricLabel, action, err == nil)
+
+	return err
+}
+
+func (k *K8sWatcher) onDeleteCIDRGroup(
+	cidrGroupName string,
+	cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup,
+	cnpCache map[resource.Key]*types.SlimCNP,
+	cs client.Clientset,
+	apiGroup, metricLabel string,
+) error {
+	delete(cidrGroupCache, cidrGroupName)
+
+	err := k.updateCIDRGroupRefPolicies(cidrGroupName, cidrGroupCache, cnpCache, cs)
+
+	k.K8sEventProcessed(metricLabel, resources.MetricDelete, err == nil)
+	k.K8sEventReceived(apiGroup, metricLabel, resources.MetricDelete, true, true)
```

Comment:
Actually, it is the same logic we are already using for CNP/CCNP: [[1]](https://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/cilium_network_policy.go#L140), [[2]](https://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/cilium_network_policy.go#L173) and [[3]](https://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/cilium_network_policy.go#L188-L189). Looking at the existent code, it is possible to have errors while updating or deleting CNPs, too.

I agree that it is questionable, but maybe it is like that because the initial intention was to report info about the event reception and not its processing, which is reported by `k.K8sEventProcessed()`?
Maybe we should reconsider the way we are using the labels for all of those metrics (CNP, CCNP and CCG) in a follow up PR?

Reply:
Opened a related issue [here](https://github.com/cilium/cilium/issues/24873)